keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,491 @@
+- Usability:
+  - Use of the secret in other resource will not be obvious to users. Therefore, users might not easily understand why the secret is not deleted.
+  - Users might need to force delete the secret on deletion or would like to avoid protection for certain secrets that already exist or that are newly created.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+There is a corner case when Reclaim Policy is Retain and PV is referencing secrets in a certain user namespace.
+In such a case, when PVC is deleted, the PV referencing secrets remains as a resource not managed by the namespace, as a result, deletion of the secrets is blocked.
+However, the user of the namespace won't notice why it is blocked and won't find a way to delete the secrets other than manually deleting finalizer.
+In addition, to make things worse, provisioner secrets that are defined by using template with PVC information,
+like "${pvc.name}", would be deleted in above case, due to the failure in finding a reference to the secret,
+because the reference to provisioner secret is not stored in PV and needs to be resolved on deletion time with PVC information that is already deleted.
+This behavior is inconsistent with other secrets, like node publish secret whose reference is stored in PV.
+
+## Design Details
+
+This feature can be implemented in the same way as `pv-protection/pvc-protection`.
+Protection logics for in-tree resources and out-of-tree resources are separeted and independently work.
+It is due to the restriction that CRDs can't be handled from in-tree controller.
+
+- In-tree resources(`Pod` and `PersistentVolume`):
+  - The deletion is blocked by using newly introduced `kubernetes.io/secret-protection` finalizer,
+  - The `kubernetes.io/secret-protection` finalizer will be always added on creation of the secret by using admission controller for in-tree resources,
+  - After the secret is requested to be deleted (`deletionTimestamp` is set), the `kubernetes.io/secret-protection` finalizer will be deleted by newly introduced in-tree `secret-protection-controller` by checking whether the secret is in-use, on every change(Create/Update/Delete) events for secrets and related resources.
+  - If the `SecretInUseProtection` feature gate is disalbed, finalizer is added on creation, but deleted regardless of whether it is in-use after `deletionTimestamp` is set. This will allow users to avoid manually deleting finalizers on the downgrading by disabling the feature.
```

Comment:
That means that in alpha, you will be:
 - still adding the finalizer at admission time
 - always removing the finalizer without checking for in-use logic

I don't think we can have that level of intrusiveness for an alpha feature.

I guess that "disabling" is not just the feature gate in this case, but rather disabling the admission plugin and the controller as well. But there is some sequencing needed here to get the clean up done properly.

In the enablement/disablement section, I think you need to detail this very explicitly. Enablement/disablement and upgrade/downgrade are the only really critical PRR things we need for an alpha feature. You need to be very explicity, like:

To enable:
 - add the admission plugin
 - enable the controller
 - enable the feature gate in kube api server

To disable:
 - first disable feature gate, then let it clean up, then disable admission plugin, then disable controller (with whatever appropriate steps need to be in place between each of these to prevent any stuck secrets)

Or whatever the actual sequence is that you are recommending. We need to be sure that:

1) By default, someone upgrading does not end up with any of this code running.
2) Users know exactly what to do to enable it.
3) Users know exactly what to do to disable it, without leaving any stuck secrets, and so that once it is fully disabled, we are back to the state where none of this code is running.
4) How to rollback a version if this has been enabled. That may be having to go through the entire disablement cycle first, to clean things up, with manual clean up as a very last resort.

Reply:
@johnbelamaric Updated. PTAL

Actually, I found that this is related to this [comment](https://github.com/kubernetes/enhancements/pull/2640#discussion_r631956311). Should we change the specification like the comment or keep this behavior?
By changing the behavior, we can provide users with a way for out-of-tree controller to avoid manually delete finalizer, like "otherwise" case of in-tree controller. 

@xing-yang @liggitt  WDYT?

Reply:
By above comment, I mean whether we should change the spec to like "Add a nob to disable out-of-tree controller" commit.